server/app/controllers/admin/AdminProgramStatusesController.java

```diff
@@ -34,6 +57,209 @@ public Result index(Http.Request request, long programId) throws ProgramNotFound
     if (!statusTrackingEnabled) {
       return notFound("status tracking is not enabled");
     }
-    return ok(statusesView.render(service.getProgramDefinition(programId)));
+    requestChecker.throwIfProgramNotDraft(programId);
+    ProgramDefinition program = service.getProgramDefinition(programId);
+    int previousStatusCount = program.statusDefinitions().getStatuses().size();
+
+    Optional<Form<ProgramStatusesEditForm>> maybeEditForm = Optional.empty();
+    if (request.method().equalsIgnoreCase(HttpVerbs.POST)) {
+      Pair<Form<ProgramStatusesEditForm>, Optional<StatusDefinitions>> validated;
+      try {
+        validated = validateEditRequest(request, program);
+      } catch (MissingStatusException e) {
+        return redirect(routes.AdminProgramStatusesController.index(programId).url())
+            .flashing(
+                "error",
+                "The status being edited no longer exists and may have been modified in a separate"
+                    + " window.");
+      }
+      maybeEditForm = Optional.of(validated.getLeft());
+      if (!validated.getLeft().hasErrors() && validated.getRight().isPresent()) {
```

Comment:
Unpack left/right asap and use them for readability.  Pair<> is usually discouraged as it lacks readability, the preference is typically a private AutoValue, but here unpacking it into readable left/right may work.

Reply:
Good suggestion for readability. My general preference is discouraging Pair<> for external APIs and that it's fine to use for internal private methods (assuming the pair isn't of the same type). If we desire AutoValue, I can look into that, but I'm not sure it provides a lot of benefit here.

I'd love it if there were easier lighter-weight ways to do this in Java.